### PR TITLE
Added SenderMail and added captcha

### PR DESCRIPTION
**Pull request by Kvolotc**

### CODE REVIEW

File: persistence/src/main/java/com/softserve/osbb/model/User.java [hunk 1]
```diff
@@ -295,7 +295,7 @@ public void setOptions(Collection<Option> options) {
         this.options = options;
     }
     
-    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
+    @ManyToOne(fetch = FetchType.EAGER, cascade = CascadeType.PERSIST)
```
Comment (KostyantynPanchenko) [inline persistence/src/main/java/com/softserve/osbb/model/User.java:298]: Why eager?

File: service/src/main/java/com/softserve/osbb/service/RegistrationService.java [hunk 2]
```diff
@@ -9,7 +9,10 @@
  */
 @Service
 public interface RegistrationService {
+	
+	StringBuilder getPassword();
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/RegistrationService.java]: Not good.

Password should be of type String. If you are constructing password with StringBuilder - just call toString at the end.

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java [hunk 3]
```diff
@@ -28,13 +35,61 @@
     private SettingsService settingsService;
     @Autowired
     private RoleService roleService;
+    
+    private  String[] alphabet = {"a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z"};
+	private Random  random = new Random();
+	private StringBuilder password;
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java]: Password should be String.

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java [hunk 4]
```diff
@@ -56,6 +111,15 @@ public Osbb registrate(Osbb newOsbb) {
         creator.setRole(roleService.getRole("ROLE_MANAGER"));
         userService.update(creator);
         return newOsbb;
-
     }
+    
+    @Override
+	public StringBuilder getPassword() {
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java]: Should be String

File: web/src/main/java/com/softserve/osbb/controller/MailController.java [hunk 5]
```diff
@@ -29,19 +31,21 @@
...
     @RequestMapping(method = RequestMethod.POST)
-    public void sendMail(@RequestBody Mail mail){
+    public ResponseEntity<Mail> sendMail(@RequestBody Mail mail){
         if (mail == null) {
             logger.warn("void mail");
         }
         
         logger.info("sending mail to "+ mail.getTo());
         
         try {
```
Comment (KostyantynPanchenko) [inline web/src/main/java/com/softserve/osbb/controller/MailController.java:41]: Refactor try-catch to single return statement

File: web/src/main/java/com/softserve/osbb/controller/MailController.java [hunk 6]
```diff
@@ -29,19 +31,21 @@
...
-            sender.send(mail.getTo(), mail.getSubject(), mail.getText());
+            sender.send( mail.getTo(), mail.getSubject(), mail.getText());
+            return new ResponseEntity<>(new Mail(mail.getTo(),mail.getText()), HttpStatus.OK);
         } catch (MessagingException e) {
             logger.error("cannot send message to" + mail.getTo());
```
Comment (KostyantynPanchenko) [inline web/src/main/java/com/softserve/osbb/controller/MailController.java:44]: Why so much logging?

File: web/src/main/java/com/softserve/osbb/controller/RegistrationController.java [hunk 7]
```diff
@@ -57,13 +57,15 @@
...
         User foundUser = userService.findUserByEmail(userRegistrationDTO.getEmail());
-        
         if (foundUser != null) {
             throw new UserAlreadyExistsException("user already exists");
         }
         
         User registeredUser = registrationService.registrate(
                 userRegistrationDTOMapper.mapDTOToEntity(userRegistrationDTO));
+        if(registrationService.getPassword() != null) {
+        	registeredUser.setPassword(registrationService.getPassword().toString());
```
Comment (KostyantynPanchenko) [inline web/src/main/java/com/softserve/osbb/controller/RegistrationController.java]: Refactor password to String

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 8]
```diff
@@ -44,18 +47,20 @@ export class UsersComponent implements OnInit {
     private houses: Array<string> = [];
     private apartment: Array<string> = [];
     genders = [
-        'male',
-        'female'
+        'Чоловік',
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts:51]: Wrong. Should be an entity and retrieved from DB.
Internationalization also should work

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 9]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
             }
         )
     }
+    
+    sendEmailActive(user:User) {
+        this.mail.to = user.email;
+        this.mail.text = 'Добрий день '+user.firstName+' '+user.lastName+
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 10]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        ' ваш акаунт був підтвердженний головою Осбб';
+        this.mail.subject = 'Активація';
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 11]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        this.mailService.sendEmail(this.mail)
+        .subscribe((data)=> {
+        });
+    }
+
+    sendEmailDeactive(user:User) {
+        this.mail.to = user.email;
+        this.mail.text = 'Добрий день '+user.firstName+' '+user.lastName+
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 12]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        ' ваш акаунт був деактивований головою Осбб';
+        this.mail.subject = 'Деактивація';
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 13]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        this.mailService.sendEmail(this.mail)
+        .subscribe((data)=> {
+        });
+    }
+
+    sendEmailRandomPassword(user:UserRegistration) {
+        console.log('User password '+user.password);
+        this.mail.to = user.email;
+        this.mail.text = 'Добрий день '+user.firstName+' '+user.lastName+
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 14]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        ' ваш акаунт зареєстрував голова Осбб , ваш логін  ' 
+        +user.email+ ' , ваш пароль до нашого сайту '+user.password+' будь-ласка не кажіть свій пароль нікому. Хорошого дня';
+        this.mail.subject = 'Реєстрація Головою Осбб';
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 15]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        this.mailService.sendEmail(this.mail)
+        .subscribe((data)=> {
+        });
+    }
+
+    sendEmailDelete(user:User) {
+        this.mail.to = user.email;
+        this.mail.text = 'Добрий день '+user.firstName+' '+user.lastName+
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/admin/components/users/users.component.ts [hunk 16]
```diff
@@ -89,6 +98,48 @@ export class UsersComponent implements OnInit {
...
+        ' ваш акаунт був видаленний головою Осбб';
+        this.mail.subject = 'Видалення';
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/admin/components/users/users.component.ts]: i18n

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 17]
```diff
@@ -99,6 +101,7 @@ export class RegistrationComponent implements OnInit {
         this.listAllOsbb();
         this.ListAllRegion();
         this.IsRegistered = true;
+        this.avtoGeneratePassword();
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: avto => auto

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 18]
```diff
@@ -119,14 +122,24 @@ export class RegistrationComponent implements OnInit {
     }
 
     onSubmitJoin() {
+    	if(this.captchaUser == null || this.captchaUser == '' ) {
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: add space after if

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 19]
```diff
@@ -119,14 +122,24 @@ export class RegistrationComponent implements OnInit {
...
+            this._toasterService.pop('error',' Ви не заповнили капчу');
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: i18n

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 20]
```diff
@@ -119,14 +122,24 @@ export class RegistrationComponent implements OnInit {
...
+            return;
+        }
+        if(this.captcha != this.captchaUser) {
+            this._toasterService.pop('error',' Ви погано заповнили капчу');
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: i18n

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 21]
```diff
@@ -119,14 +122,24 @@ export class RegistrationComponent implements OnInit {
...
         this.SenderJoin();
     }
 
     SenderJoin(): any {
         let isSuccessful: boolean = false;
         this.registerService.registerUser(this.newUser)
             .subscribe(
-                data => {                
+                data => {
+                	this.sendEmailToUser();
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: Can you refactor sentEmailToUser and sendEmailToGolova to single method sendTo(<User || String> to)?

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 22]
```diff
@@ -474,4 +476,62 @@ export class RegistrationComponent implements OnInit {
         console.log('error msg' + error)
     }
     
+    sendEmailToUser() {
+        this.mail.to = this.newUser.email;
+        this.mail.text = 'Добрий день '+this.newUser.firstName+' '+this.newUser.lastName+
+        ' дякуємо за реєстрацію на сайті OSBB , ваш акаунт буде розлянуто головою ОСББ , після підтвердження головою ОСББ вам на пошту прийде підтвердження';
+        this.mail.subject = 'Регістрація';
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: Реєстрація

File: web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts [hunk 23]
```diff
@@ -474,4 +476,62 @@ export class RegistrationComponent implements OnInit {
         console.log('error msg' + error)
     }
     
+    sendEmailToUser() {
```
Comment (KostyantynPanchenko) [inline web/src/main/resources/public/src/app/registration/registration_user/registration.component.ts]: refactor to single method sendTo(params)

**Review by KostyantynPanchenko — CHANGES_REQUESTED** [review]
Please make changes / provide feedback

File: persistence/src/main/java/com/softserve/osbb/model/User.java (hunk 1 above)
Comment (Kvolotc) [reply persistence/src/main/java/com/softserve/osbb/model/User.java:298]: it was that befor me . for @ManyToOne it is default

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java [hunk 24]
```diff
@@ -28,13 +35,60 @@
     private SettingsService settingsService;
     @Autowired
     private RoleService roleService;
+    
+    private  String[] alphabet = {"a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z"};
+	private Random  random = new Random();
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java:44]: Indentation lines 40 -83

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java [hunk 25]
```diff
@@ -28,13 +35,60 @@
     private SettingsService settingsService;
     @Autowired
     private RoleService roleService;
+
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java:42]: Add blank lines before annotations

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java [hunk 26]
```diff
@@ -28,13 +35,60 @@
...
+                password.append(alphabet[random.nextInt(alphabet.length)]);
+            }
+            else if(randThreshold % DIVIDER == INDEX_OF_UPPERCASE_LETTER) {
+                password.append(alphabet[random.nextInt(alphabet.length)].toUpperCase());
+            }
+            else{
+                password.append(random.nextInt(RANDOM_AREA));
+            }
+        }
+       return (this.password = password.toString());
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java:86]: This is assignment or checking for equality?

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java (hunk 24 above)
Comment (Kvolotc) [reply service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java:44]: here all right

File: service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java (hunk 26 above)
Comment (Kvolotc) [reply service/src/main/java/com/softserve/osbb/service/impl/RegistrationServiceImpl.java:86]: for Initialization  password and than return

**Review by KostyantynPanchenko — APPROVED** [review]

**Review by aprudyus — APPROVED** [review]

**Review by OlegKykena — APPROVED** [review]

**Pull request merged**